File: src/projekat/PAMformat.java
package projekat;


import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PAMformat extends Format {

    private int depth = 0;
    private int maxval = 0;
    private String tupl_type = "";

    private String widths;
    private String heights;
    private String maxvals;


    public PAMformat(String directory) {
        super(directory);
    }

    @Override
    public void parse(ArrayList<Pixel> pikseli) {
        /*FileReader file = null;
        try {
            file = new FileReader(directory);
        } catch (FileNotFoundException e) {
           // throw GreskaOtvaranjeDatoteke();
        }
        BufferedReader bf = new BufferedReader(file);
*/

        String line;
        int size_info = 0;

        Pattern rx1 = Pattern.compile("([A-Z]*) ([0-9]*)");
        Pattern rx2 = Pattern.compile("([A-Z]*) ([A-Z]*)");


       // Stream<String> lines = bf.lines();

        FileInputStream fileInputStream = null;
        try {
            fileInputStream = new FileInputStream(directory);
        } catch (FileNotFoundException e) {
            new Greska("Fajl nije pronadjen!");
        }
        Scanner scan = new Scanner(fileInputStream);


        int cnt = 0;

        while(cnt != 6) {
            line = scan.nextLine();
            Matcher matcher1 = rx1.matcher(line);
            Matcher matcher2 = rx2.matcher(line);

            if(cnt == 0){

            }
            if (matcher1.matches()) {
                switch (cnt) {
                    case 1: {
                        width = Integer.parseInt(matcher1.group(2));
                        widths = Integer.toString(width, 10);
                        break;
                    }
                    case 2: {
                        height = Integer.parseInt(matcher1.group(2));
                        heights = Integer.toString(height, 10);
                        break;
                    }
                    case 3: {
                        depth = Integer.parseInt(matcher1.group(2));
                        break;
                    }
                    case 4: {
                        maxval = Integer.parseInt(matcher1.group(2));
                        maxvals = Integer.toString(maxval, 10);
                        break;
                    }
                }
            } else {
                if (matcher2.matches()) {
                    tupl_type = matcher2.group(2);
                }
                else tupl_type = "RGB_ALPHA";
            }
            cnt++;
        }
        if (tupl_type == "RGB_ALPHA") {
            size = 4 * width * height;
        }
        else size = 3 * width * height;


        try {
            fileInputStream = new FileInputStream(directory);
        } catch (FileNotFoundException e) {

        }
        DataInputStream dis = new DataInputStream(fileInputStream);
        char c;
        int headerLines = 7;
        while(headerLines > 0) {
            do {
                try {
                    c = (char) (dis.readUnsignedByte());
                } catch (IOException e) {
                    c = ' ';
                }
            } while (c != '\n');
            headerLines--;
        }

        for(int i = 0; i < size; i++){
            try {
                int b = dis.readUnsignedByte();
                data.add(b);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }



        if (this.tupl_type == "RGB_ALPHA") {
            for (int i = 0; i < size; i += 4) {
                Pixel p = new Pixel(data.get(i), data.get(i+1), data.get(i+2), data.get(i+3));
                pikseli.add(p);
            }
        }
        else {
            for (int i = 0; i < size; i += 3) {
                Pixel p = new Pixel(data.get(i), data.get(i+1), data.get(i+2), 255);
                pikseli.add(p);
            }
            this.size = 4 * width * height;
        }

        try {
            fileInputStream.close();
        } catch (IOException e) {}

    }


    @Override
    public void eksportuj(String directory) {
        try {
            FileWriter fileWriter = new FileWriter("Data/" + directory, StandardCharsets.ISO_8859_1);
            fileWriter.write("P7\nWIDTH ");

            widths = Integer.toString(width, 10);
            fileWriter.write(widths);
            fileWriter.write("\n");
            fileWriter.write("HEIGHT ");
            heights = Integer.toString(height, 10);
            fileWriter.write(heights);
            fileWriter.write("\n");
            fileWriter.write("DEPTH ");
            fileWriter.write("4");
            fileWriter.write("\n");
            fileWriter.write("MAXVAL ");
            fileWriter.write("255");
            fileWriter.write("\n");
            fileWriter.write("TUPLTYPE ");
            fileWriter.write("RGB_ALPHA");
            fileWriter.write("\n");
            fileWriter.write("ENDHDR");
            fileWriter.write("\n");

            data.forEach(c->{
                try {
                    int i = c;
                    fileWriter.write(i);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            });

            fileWriter.close();
        } catch (FileNotFoundException e) {
            new Greska("Fajl vec postoji!");
        } catch (IOException e) {

        }
    }

    public void setWidth(int w) { this.width = w; widths = Integer.toString(width,10);}
    public void setHeight(int h) { this.height = h; heights = Integer.toString(height,10);}
    public void setTuplTypeRGBA() { this.tupl_type = "RGB_ALPHA"; }
    public void setAll(int wid, int hei, int si, ArrayList<Integer> d){
        this.setWidth(wid);
        this.setHeight(hei);
        this.setSize(si);
        this.setData(d);
    }

    public String getTuplType()  { return this.tupl_type; }
}

File: c++ project/ProjekatPOOP/Project/Project/PAMformat.cpp
#include "PAMformat.h"

void PAMformat::parse(vector<Pixel*>& pikseli)
{
	ifstream file(directory);
	if (file.is_open() == false) {
		throw GreskaOtvaranjeDatoteke();
	}
	else {
		string line;
		int cnt = 0;
		int size_info = 0;

		regex rx1("([A-Z]*) ([0-9]*)");
		regex rx2("([A-Z]*) ([A-Z]*)");

		while (getline(file, line) && cnt != 6) {
			switch (cnt) {
			case 0: {
				if (line != "P7") throw GreskaZapisDatoteke();
				break;
			}
			case 1: {
				smatch result;
				if (regex_match(line, result, rx1)) {
					width = atoi(result.str(2).c_str());
					itoa(width, widths, 10);
				}
				break;
			}
			case 2: {
				smatch result;
				if (regex_match(line, result, rx1)) {
					height = atoi(result.str(2).c_str());
					itoa(height, heights, 10);
				}
				break;
			}
			case 3: {
				smatch result;
				if (regex_match(line, result, rx1)) {
					depth = atoi(result.str(2).c_str());
				}
				break;
			}
			case 4: {
				smatch result;
				if (regex_match(line, result, rx1)) {
					maxval = atoi(result.str(2).c_str());
					itoa(maxval, maxvals, 10);
				}
				break;
			}
			case 5: {
				smatch result;
				if (regex_match(line, result, rx2)) {
					tupl_type = result.str(2);
				}
				else tupl_type = "RGB_ALPHA";
				break;
			}
			case 6: break;
			}

			cnt++;
		}


		if (tupl_type == "RGB_ALPHA") {
			size = 4 * width * height;
		}
		else size = 3 * width * height;

		/*for (int i = 0; i < size; i++) {
			getline(file, line);
			for (auto i : line) data.push_back(i);
			i += line.size();
			if(i != size - 1) data.push_back('\n');
			else {

			}

		}*/

		/*while (file.eof() != true) {
			getline(file, line);
			for (auto i : line) data.push_back(i);
			data.push_back('\n');
		}*/

		vector<char> pom;
		pom.resize(size);

		file.read(pom.data(), pom.size());
		data.assign(pom.begin(), pom.end());

		if (this->tupl_type == "RGB_ALPHA") {
			for (int i = 0; i < size; i += 4) {
				Pixel* p = new Pixel(data[i], data[i + 1], data[i + 2], data[i + 3]);
				pikseli.push_back(p);
			}
		}
		else {
			for (int i = 0; i < size; i += 3) {
				Pixel* p = new Pixel(data[i], data[i + 1], data[i + 2], 255);
				pikseli.push_back(p);
			}
			this->size = 4 * width * height;
		}
	}
}

void PAMformat::exportuj(string naziv)
{
	FILE* file = fopen(naziv.c_str(), "wb");
	if (file == nullptr) throw GreskaOtvaranjeDatoteke();

	fprintf(file, "P7\nWIDTH ");
	itoa(width, widths, 10);
	fprintf(file, widths);
	fprintf(file, "\n");
	fprintf(file, "HEIGHT ");
	itoa(height, heights, 10);
	fprintf(file, heights);
	fprintf(file, "\n");
	fprintf(file, "DEPTH ");
	fprintf(file, "4");
	fprintf(file, "\n");
	fprintf(file, "MAXVAL ");
	fprintf(file, "255");
	fprintf(file, "\n");
	fprintf(file, "TUPLTYPE ");
	fprintf(file, "RGB_ALPHA");
	fprintf(file, "\n");
	fprintf(file, "ENDHDR");
	fprintf(file, "\n");

	for (auto& i : data) {
		fputc(i, file);
	}

	fclose(file);
}

void PAMformat::setAll(int wid, int hei, int si, vector<unsigned char> d)
{
	this->setWidth(wid);
	this->setHeight(hei);
	this->setSize(si);
	this->setData(d);
}
